first draft of workflow format (#171)
OCR-D/spec: workflow-format.md

# OCR-D Workflow format
> Sequence of [processor](glossary#processor) [calls](cli)

Revision: 1

## Rationale

OCR consists of many [steps](glossary#activities) that need to be run in
sequence ([*workflow*](glossary#workflow)). In OCR-D, these steps are

> implemented in [*processors*](https://ocr-d.de/en/workflows), executables that

Reviewer: ```suggestion
implemented in [*processors*](glossary#ocr-d-processor), executables that
```

read and write files to one or more `mets:fileGrp` in a METS document. While it
is possible to define such a workflow as a sequence of commands in a shell
script, that approach has many drawbacks (no checks for correct input/output
wiring, no syntax checks, easy to end up with an invalid
[workspace](glossary#workspace)).

A subset of shell script is however a good solution to define a common,
workflow-engine-agnostic syntax for workflows in OCR-D. It is straightforward
to tokenize [according to
POSIX](https://pubs.opengroup.org/onlinepubs/009695399/utilities/xcu_chap02.html#tag_02_03)
(`sh`) and every processor invocation can be parsed according to the [OCR-D CLI
spec](cli). Every step can be reproduced easily by executing it in a command
shell.

While such an OCR-D workflow description "script" (*OCRD-WF*) is syntactically
compatible with a POSIX-compliant command shell, it has different semantics and
supports only a small subset of shell script syntax and functionality.

## File structure and syntax

> A OCRD-WF script has two logical sections: preamble and steps.

Reviewer 2: ```suggestion
An OCRD-WF script has two logical sections: preamble and steps.
```

The preamble contains the shebang and any variable
assignments while the steps contain the processor calls. Both preamble and steps can
contain any number of comment lines.

### Shebang

The first line of every OCRD-WF must match the regular expression

> `#!/usr/bin/env ocrd-wf-v[1-9]+[0-9]*\n` 

Reviewer 2: ```suggestion
`#!/usr/bin/env ocrd-wf-v[1-9][0-9]*\n` 
```

(including zero) followed by the character "#" (*pound sign*, `\x23`).

### Line continuation

If a line ends with "\" (*backslash*, `\x5c`), backslash and newline are
removed and the next line, without leading whitespace, is appended. This is
repeated as long as the next line ends with a backslash.

### Variable Assignments

An OCRD-WF can contain variable assignments of the form `name=value\n`. `name` must be
a valid `sh` identifier. Everything to the right of `=` is parsed as a string after expanding quotation (i.e. removing enclosing pairs of double `"` or single `'` quotes, or backslash `\` to _escape_ the special meaning of the following character, with `\\`, `\"`, and `\'` for literal backslash, double quote, and single quote, respectively). Note: whitespace is not allowed anywhere, except as quoted part of `value`.

<!-- TODO define behavior? -->
**NOTE** The semantics of assigning `value` to `name` are implementation-dependent

## Parsing Algorithm

The parsing algorithm consists of a set of passes. It should inform
implementers, but does not require an exact implementation, as long as
the result is the same.

### Check shebang

> * assert that line[0] is `#!/usr/bin/env ocrd-wf`

Reviewer 2: Missing version after `ocrd-wf`?

    * split line into tokens according to POSIX rules
    * if there is more than one token:
      * raise an error "tokens after assignment"
    * add first token to assignments_tokens
    * continue
  * raise an error "unhandled line"
* return steps_tokens, assignments_tokens

### Parse

* set steps to empty list
* for step_tokens in steps_tokens:
  * for tokens in step_tokens:
    * set step to empty map
    * set step.parameters to empty map
    * set step.executable to tokens[0]
    * remove tokens[0]
    * parse tokens to step [according to OCR-D CLI spec](cli) but raise an error for these options:
      * `-m/--mets` (provided by the workflow engine)
      * `-h/--help`
      * `--version`
      * `-J--dump-json`
    * add step to steps
* set assignments to empty map
* for token in assignments_tokens:
  * split tokens at the first equal sign into key and value
  * set assignments.key to value
* return steps, assignments

## Validation

### Well-formedness

OCRD-WF is *well-formed* if it can be parsed.

### Resolveability

OCRD-WF is *resolveable* if it is well-formed and
  * all executables are available (either as an executable in `$PATH` or some abstraction of it)

>   * all exceutbales can be resolved to their resp. [`ocrd-tool.json`](ocrd_tool) (e.g. [`--dump-json`](cli#--J--dump-json))

Reviewer: ```suggestion
  * all executables can be resolved to their resp. [`ocrd-tool.json`](ocrd_tool) (e.g. [`--dump-json`](cli#--J--dump-json))
```

  * the parameters in all steps are valid according to their JSON-Schema definition.

Resolveability can only be determined in the context of a deployment.

### Consistency

OCRD-WF is *consistent* if it is resolveable and all input file groups
  * are either output file groups of previous steps or

>   * exist in the METS file this workflow is executed on

Reviewer: ```suggestion
  * are present in the METS file this workflow is executed on
```

### EBNF

<!-- TODO: questionable whether helpful -->

```c
nl = "\x0a"
ws = "\x09" | "\x20"
any = ("\x00" .. "\xffff") - nl
bslash = "\x5c"
not-bslash = any - bslash
alpha = "A" .. "Z" | "a" .. "z"
alnum = alpha | "0" .. "9"
comment-prefix = "#"

> shebang = "#!/usr/bin/env ocrd-wf\n"

Reviewer 2: Missing version?

# OCR-D Workflow format
> Sequence of [processor](glossary#processor) [calls](cli)

Revision: 1

## Rationale

OCR consists of many [steps](glossary#activities) that need to be run in
sequence ([*workflow*](glossary#workflow)). In OCR-D, these steps are

> implemented in [*procesors*](https://ocr-d.de/en/workflows), executables that

Reviewer: proces**s**ors

supports only a small subset of shell script syntax and functionality.

## File structure and syntax

A OCRD-WF script has two logical sections: preamble and steps.

The preamble contains the shebang and any variable
assignments while the steps contain the processor calls. Both preamble and steps can
contain any number of comment lines.

### Shebang

The first line of every OCRD-WF must match the regular expression
`#!/usr/bin/env ocrd-wf-v[1-9]+[0-9]*\n` 

The number represents the revision of this spec, so initially, all OCRD-WF
should begin with `#!/usr/bin/env ocrd-wf-v1`.

### Comments

Everyline that begins with any whitespace followed by the character "#" (*pound
sign*, `\x23`).

### Line continuation

If a line ends with "\" (*backslash*, `\x5c`), backslash and newline are
removed and the next line, without leading whitespace, is appended. This is
repeated while the joined ends with a backslash.

### Variable Assignments

An OCRD-WF can contain variable assignments of the form `name=value\n`. `name` must be
a valid `sh` identifier, everything `=` is parsed as a string.

<!-- TODO define behavior? -->

> **NOTE** The semantics of assigning `value` to `name` are implementation-dependant

Reviewer 3: In particular, where and how in the step definition do you _use_ variables? You'll have to describe variable expansion really, esp. in its interaction with quote expansion. In practice, we'd need to be able to do things like:
```sh
bin=wolf
param_bin='{"input_feature_filter": "binarized", "output_feature_added": "binarized", "command": "scribo-cli $bin @INFILE @OUTFILE --enable-negate-output --k 0.2 --win-size 301"}'
ocrd-preprocess-image         -I OCR-D-IMG               -O OCR-D-BIN-$bin          -P level-of-operation page -p "$param_bin"
ocrd-cis-ocropy-denoise       -I OCR-D-BIN-$bin          -O OCR-D-BIN-$bin-DEN      -P level-of-operation page
ocrd-tesserocr-deskew         -I OCR-D-BIN-$bin-DEN      -O OCR-D-BIN-$bin-DEN-FLIP -P operation_level page
ocrd-cis-ocropy-deskew        -I OCR-D-BIN-$bin-DEN-FLIP -O OCR-D-BIN-$bin-DEN-DESK -P level-of-operation page
ocrd-tesserocr-segment-region -I OCR-D-BIN-$bin-DEN-DESK -O OCR-D-SEG-REG
ocrd-preprocess-image         -I OCR-D-SEG-REG           -O OCR-D-SEG-REG-BIN-$bin  -P level-of-operation region -p "$param_bin"
...
ocrd-preprocess-image         -I OCR-D-SEG-LINE          -O OCR-D-SEG-LINE-BIN-$bin -P level-of-operation line -p "$param_bin"
```

(We may need variable expansion in fileGrp names whenever multiple alternatives are used for different purposes, like different binarizations for segmentation and recognition. In the parameter JSON and parameter literal options, the need is obvious – for brevity and changability.)

> **NOTE** The semantics of assigning `value` to `name` are implementation-dependant

Author: For this initial draft, I side-stepped the question how variables come into play, because I need input (like yours, thx) on how this should work.

I really like how standard `=` variable assignments work in Makefiles and would like to see the same semantics with `/bin/sh` syntax. I.e. the example you posted would be valid but at workflow execution time, users should be able to *override* those assignments to use a different model, a different set of parameters etc.

*consistency* would then also include a check that there is an assignment (either in the file or provided by the workflow engine) to replace all variables in the file.

For ease of parsing and visual effect, I'd prefer `${foo}` syntax over `$foo`.

Also, I think this should not adhere to bourne shell variable expansion (which e.g. does not work inside single quotes and can be escaped with a backslash) but rather the more aggressive `make` approach to replace all `$(var)` calls everywhere.

> **NOTE** The semantics of assigning `value` to `name` are implementation-dependant

Reviewer 3: > I really like how standard `=` variable assignments work in Makefiles and would like to see the same semantics with `/bin/sh` syntax. I.e. the example you posted would be valid but at workflow execution time, users should be able to _override_ those assignments to use a different model, a different set of parameters etc.

I agree with you aesthetically, but then you definitely need a completely different non-sh (non-static scoping) semantics. You essentially want the declarative model for makefiles. Which is not very intuitive for most users I must say.

> Also, I think this should not adhere to bourne shell variable expansion (which e.g. does not work inside single quotes and can be escaped with a backslash) but rather the more aggressive `make` approach to replace all `$(var)` calls everywhere.

The `make` approach is problematic itself though – exactly _because_ of its word expansion (and ignorance of any quotation). In sh-like syntax, you can at least have nested quotation of arbitrary depth.

Having a concise, readable and runnable declarative syntax for workflows was among my initial motivations for `workflow-configuration`. This could also be parsed to run independent of `make`. I fail to see the merit of introducing yet another syntax if not for simplicity and intuitivity (which I still believe sh-like configuration serves best).

> **NOTE** The semantics of assigning `value` to `name` are implementation-dependant

Author: > You essentially want the declarative model for makefiles. Which is not very intuitive for most users I must say.

Why do you think it is unintuitive? The OCRD-WF defines variables and command line calls. Variables are hoisted at parse-time (which is different from shell script, true, but since we do not support any conditional syntax, it makes no sense to support non-global scoping).

> In sh-like syntax, you can at least have nested quotation of arbitrary depth.

That's a good point. make's interpolation is more powerful IMHO (*because* it doesn't care about quotes) but it can lead to unintuitive behavior, I agree. 

> This could also be parsed to run independent of make.

But only if you stick to a completely static sequence of steps, without any fileGrp-guessing logic and the like - nothing against trying to be smart, just that Makefile's are not trivial to parse, with all its more or less arcane incantations :)

> I fail to see the merit of introducing yet another syntax if not for simplicity and intuitivity (which I still believe sh-like configuration serves best).

The merit of this syntax is supposed to be interoperability. At the very least, OCRD-WF should be exportable to (less intuitive than hand-written but still complete) workflow-configuration makefiles or `ocrd process` calls.

> **NOTE** The semantics of assigning `value` to `name` are implementation-dependant

Reviewer 4: I would argue that this syntax is as intuitive as it can get with OCR-D and is quite easy when you are familiar with the OCR-D syntax. Many users just want to copy and paste the example calls from our website and we can't provide examples for all processing solutions. So it would make documentation and usage way easier if you could just stick to one syntax and generate the rest from that.

> **NOTE** The semantics of assigning `value` to `name` are implementation-dependant

Reviewer 3: > > You essentially want the declarative model for makefiles. Which is not very intuitive for most users I must say.
> 
> Why do you think it is unintuitive? The OCRD-WF defines variables and command line calls. Variables are hoisted at parse-time (which is different from shell script, true, but since we do not support any conditional syntax, it makes no sense to support non-global scoping).

Because variables can get their values from below the line where they are used, and multiple assignments make no sense (and don't work). All this is unexpected for a syntax that looks (and is described as) sh-like. (Remember, I do like that kind of syntax myself, but the question is: What is best for normal users?) And since practically (because of the large amounts of writing effort and duplication needed in typical workflows) we'll rely on variables quite a lot, this is a big decision...

> > In sh-like syntax, you can at least have nested quotation of arbitrary depth.
> 
> That's a good point. make's interpolation is more powerful IMHO (_because_ it doesn't care about quotes) but it can lead to unintuitive behavior, I agree.

Not just unintuitive. `make`'s tokenization / word expansion is incapable of distinguishing between a space that's part of the value and a delimiter. For example, in `variable = value `, does the space after `=` belong to `value` or not? (Could be part of a file name, which needs to be quoted before passing to the shell.) I had problems with this in workflow-configuration left and right (e.g. for passing the parameter JSON).

> > This could also be parsed to run independent of make.
> 
> But only if you stick to a completely static sequence of steps, without any fileGrp-guessing logic and the like - nothing against trying to be smart, just that Makefile's are not trivial to parse, with all its more or less arcane incantations :)

True, this translation could only be done via `make` itself. I was thinking of something like `make -Bsn > workflow.sh` (which could be usually rewritten for `ocrd process`). But sure, that's just a form of conversion based on runtime model equivalence – not a formal model.

> The merit of this syntax is supposed to be interoperability.

Granted, that's even more important than simplicity. But a truly sh-based restricted syntax would be both interoperable and simple, even with standard backslash/quote expansion. (Remember I was arguing against make-like _dedicated_ syntax.)

How do you plan on implementing this translation anyway? Some form of `bash -x workflow.ocrd.sh` or `set -n` perhaps, or are you going to write a (restricted) sh parser of your own (talking of incantations...)?

> **NOTE** The semantics of assigning `value` to `name` are implementation-dependant

Author: > Because variables can get their values from below the line where they are used, and multiple assignments make no sense (and don't work).

True, that's what I meant by "hoisting", that variable assignments are processed before any steps, even if they assignments and steps are interwoven. This is decidedly non-sh behavior and unintuitive if you approach OCRD-WF as a shell script. But if you *do* approach it like a shell script, you might be tempted to assume statefulness and expect conditionals or loop constructs to be supported, which they are not. There absolutely is a benefit to allowing state/conditionals (like the group-guessing in workflow-configuration scripts) and if we want to support this, we should decide on it now. If not, I'd prefer the make assignment semantics.

> I was thinking of something like `make -Bsn > workflow.sh` (which could be usually rewritten for `ocrd process`)

This pattern never occured to me but is very clever. Can you add that to https://hackmd.io/emBLZdD5SyWietFxLLpnQg?

> But a truly sh-based restricted syntax would be both interoperable and simple, even with standard backslash/quote expansion.

I'm with you, I care about simplicity and interoperability more than about commenting lines with escaped backslashes :)

> How do you plan on implementing this translation anyway? Some form of bash -x workflow.ocrd.sh or set -n perhaps, or are you going to write a (restricted) sh parser of your own (talking of incantations...)?

Parsing with the algorithm outlined here results in a data structure

```yaml
assignments:
  foo: bar
  bar: 42
steps:
  - executable: ocrd-foo
    parameters:
      baz: foo
    input_file_grp: MAX
    output_file_grp: FOO
```

Then we'd have adapters in `ocrd_modelfactory` that generate `ocrd process` calls or `workflow-configuration` makefiles or the CSV that @VolkerHartmann uses for invoking taverna or a shell script (which would be more or less the same as the OCRD-WF input).

```

> **NOTE** The semantics of assigning `value` to `name` are implementation-dependant

Reviewer 3: > But if you _do_ approach it like a shell script, you might be tempted to assume statefulness and expect conditionals or loop constructs to be supported, which they are not

Yes. And statefulness is something that makes your model-driven semantics hard again. Not sure myself whether users actually _want_ statefulness. But if it's going to be the declarative, make-like syntax, then we need a better metaphor than the current `shell script` / `command shell` / `ocrd+sh` / `.ocrd.sh` formulations. The term _configuration_ language does usually not cover variable references, but _macro_ language is too much related to actual programming...

> There absolutely is a benefit to allowing state/conditionals (like the group-guessing in workflow-configuration scripts) and if we want to support this, we should decide on it now.

Yes, and – as pointed out above – that's probably true not just for static conditionals (i.e. file checks or variable references) but also dynamic switches (dependent on a previous result). But this would have to go beyond OCR-D CLI (and into _OCR-D-Q_) anyway.

If what we can offer here is too little for what actual OCR-D users need and can have with their own shell scripts or makefiles, then OCRD-WF is not going to be used much, I am afraid. Perhaps it's just too early to decide on all this.

> Can you add that to https://hackmd.io/emBLZdD5SyWietFxLLpnQg?

Done.

> Parsing with the algorithm outlined here results in a data structure

I see. So you _are_ planning on implementing your own parser. I hope this is not necessary – if we could find a language suited for us, then we could just delegate implementation and documentation to that.

> **NOTE** The semantics of assigning `value` to `name` are implementation-dependant

Reviewer 3: Common Workflow Language does have [variable references](https://www.commonwl.org/user_guide/06-params/index.html) BTW. But it's quite a complex spec itself, definitely not simple.

Do we target human readers/writers here, or will they be ultimately using some GUI in the future anyway?

> **NOTE** The semantics of assigning `value` to `name` are implementation-dependant

Author: We were considering CWL as an alternative to Taverna way back at the beginning of phase 2 of OCR-D but decided against it then because Taverna offered a GUI (IIRC, @VolkerHartmann?). I still like the design a lot, but it is very different from shell script.

> Do we target human readers/writers here, or will they be ultimately using some GUI in the future anyway?

I'd say that human-readability and intuitiveness are very important. A "tiny" language based on shell script fits that description and allows us, asap, to provide reusable workflows in a repository like https://github.com/kba/ocrd-workflows

> If what we can offer here is too little for what actual OCR-D users need and can have with their own shell scripts or makefiles, then OCRD-WF is not going to be used much, I am afraid. 

The advantage over a plain shell script would be in the tooling and validation. I agree with you that it is pointless to define a script/configuration/workflow language that is then not used. We *do want* implementers to use OCRD-WF, the common workflows defined should be the basis for integration into digitization workflow software like Kitodo, Visual Library or DWork and the upcoming HTTP API should make use of it (hence the "IANA considerations").

> Perhaps it's just too early to decide on all this.

Maybe we can discuss this in the Open Tech Call on Thursday, so I understand better what users/developers expect from such a workflow language and adapt the proposal accordingly. I think a solution (that can be revised with new requirements later) is fairly urgent but we should not prevent future use cases with early design choices.

> **NOTE** The semantics of assigning `value` to `name` are implementation-dependant

Reviewer 3: > > Yes. And statefulness is something that makes your model-driven semantics hard again. Not sure myself whether users actually _want_ statefulness. But if it's going to be the declarative, make-like syntax, then we need a better metaphor than the current `shell script` / `command shell` / `ocrd+sh` / `.ocrd.sh` formulations. The term _configuration_ language does usually not cover variable references, but _macro_ language is too much related to actual programming...

> I'd say that human-readability and intuitiveness are very important. A "tiny" language based on shell script fits that description and allows us, asap, to provide reusable workflows

Let's stay with the _restricted shell script_ (as opposed to custom declarative syntax) idea for a moment.

Is the statefulness of sh-like variables really a problem? Yes, variables can be re-assigned (before re-use). But that does not necessarily make your semantics difficult. I don't think you actually need to _keep the variable_ values after computing the commands they are used to formulate. You could basically just run the entire OCRD-WF through sh with `sh -x workflow.ocrd.sh 2>&1 | sed -ne "s/^+ //p"` to get your (machine-readable) _steps_ – without the need for a parser for variables and quotation. Whatever human-readable combination of variable assignments was needed to arrive at this does not matter in the end. It's only the expanded result that counts. (You can then parse the verbatim commands to arrive at your model.)

Arguably, as soon as you do allow dynamic workflow switches, you will need _conditionals_ though. Representing this is going to be more difficult, but does not necessitate declarative syntax either. In fact, it could still be done with a proper sh subset (`if` / `then` / `elif`/ `fi`). You loose the free parser though.

On that broader topic again – dynamic switches, depending on result _checks_ – today's call made me think about the following simple concept: How about we introduced a _dedicated CLI_ for OCR-D workflow quality estimators analogous to OCR-D workflow processors? Modules could bundle their knowledge about what a good result is for a particular processor along with everything else. And specialized modules could provide QE tools by the bunch. Let's call them _benchmarks_ for the time being. We are primarily interested in a benchmark's _score_, which needs to be compared to some _threshold_ to determine if the processor's result was "okay" enough to continue processing. That threshold could be different, depending on the workflow configuration. Benchmarks could also have configurable parameters (like model files) of their own. 

An important question is how to deal with page-wise vs document-wise quality. Do we want to stop if even a single page indicated failure? Or does it take a bad average? Or a pro-longed series of bad pages? Or a `min-worst-page` / `min-average` kind of set-up?

Regardless, besides the binary `score > threshold` check we might also be interested in additional, verbose output on what was analysed and what patterns were found – either as part of logging, or as a structured _report_. So we also must allow creating an output fileGrp.

Now let's assume a call to a benchmark looks like this:
```sh
ocrdtest-cis-ocropy-binarization -I OCR-D-BIN-WOLF -O OCR-D-BIN-WOLF-TEST-OCRO -P method component-statistics -P threshold 0.9
```
(with its return value indicating success or failure, just as the return value of a processor).

Then we could write _dynamic workflows_ like so:
```sh
if BIN=wolf
  BINARIZED=OCR-D-BIN-$BIN
  ocrd-olena-binarize -I OCR-D-IMG -O $BINARIZED -P impl $BIN
  ocrdtest-cis-ocropy-binarization -I $BINARIZED -O $BINARIZED-TEST-OCRO -P method component-statistics -P threshold 0.9
then : 
elif BIN=ocropy
  # try another algorithm
  BINARIZED=OCR-D-BIN-$BIN
  ocrd-cis-ocropy-binarize -I OCR-D-IMG -O $BINARIZED -P method $BIN -P level-of-operation page
  ocrdtest-cis-ocropy-binarization -I $BINARIZED -O $BINARIZED-TEST-OCRO -P method component-statistics -P threshold 0.9
then : 
elif BIN=wolf
  # try another parameter
  BINARIZED=OCR-D-BIN-$BIN-HIGH
  ocrd-olena-binarize -I OCR-D-IMG -O $BINARIZED -P impl $BIN -P k 0.1
  ocrdtest-cis-ocropy-binarization -I $BINARIZED -O $BINARIZED-TEST-OCRO -P method component-statistics -P threshold 0.9
then : 
else 
  # give up the workflow
  exit 1
fi
ocrd-tesserocr-deskew -I $BINARIZED -O OCR-D-DESK -P operation_level page
if CROPPED=OCR-D-CROP-TESS
  ocrd-tesserocr-crop -I OCR-D-DESK -O $CROPPED
  ocrdtest-leptonica-cropping -I $CROPPED -P threshold 0.7
then : 
elif CROPPED=OCR-D-CROP-ANY
  ocrd-anybaseocr-crop -I OCR-D-DESK -O $CROPPED
  ocrdtest-leptonica-cropping -I $CROPPED -P threshold 0.5
then : 
else 
  # omit cropping
  CROPPED=OCR-D-DESK
fi
ocrd-tesserocr-segment-region -I $CROPPED -O OCR-D-SEG-TESS 
# and so on...
```
(borrowing the `sh -e` convention that any non-zero return value causes the workflow to fail, except if it was immediate to a conditional expression)

That would be fully dynamic, but still not allow arbitrary information flow like determining model names. For the latter, some notion of _functional_ CLI would be needed...

## Parsing Algorithm

The parsing algorithm consists of a set of passes. It should inform
implementers but does not require require an exact implementation, as long as
the result is the same.

**NOTE** Unlike in POSIX shell, comments are removed before line continuations are
resolved, so you can intersperse line continuation with comments:

```sh
ocrd-foo \
  # This parameter does bar
  -P foo bar

> ```

Author: It's a pet peeve of mine TBH that you cannot comment on continued lines in shell script. I understand that this is not POSIXly correct but how does it make implementation harder? If the first step in an implementation consists of removing empty lines and comments - and we allow only full-line comments - I think it makes implementation actually simpler.

I agree that it's probably a bad design decision to allow syntax that will trip up `/bin/sh` if you try to run an OCRD-WF as a shell script. But should you do that?

> ```

Reviewer 3: > but how does it make implementation harder? If the first step in an implementation consists of removing empty lines and comments - and we allow only full-line comments - I think it makes implementation actually simpler.

It would interfere with the other provision about quoted newlines:
```sh
# a very long comment indeed which \
  needs to be continued (which is covered by POSIX)
echo first command
```
If you blindly strip away the first line, then the second becomes invalid. You could still try to do your own thing by first expanding backslash-newlines and then removing comments. But you cannot rely on standard implementations anymore.

(And I don't think it's a good idea to forbid continued comment lines, because its yet another nonstandard/unexpected syntax.)

> and is unusual/unexpected (not only to shell-experienced users but also editors that understand shell syntax and can highlight it).

Any comment on that aspect?

> I agree that it's probably a bad design decision to allow syntax that will trip up `/bin/sh` if you try to run an OCRD-WF as a shell script. But should you do that?

No, I wouldn't really care about that problem.

> ```

Author: > and is unusual/unexpected (not only to shell-experienced users but also editors that understand shell syntax and can highlight it).

Neither GitHub nor Vim correctly highlight comment continuations. POSIXly correct but not correctly implemented. Why would use those over just prepending `#` to the continued line?

```sh
# comment \
  continued
ls
```

However neither GH nor Vim has a problem with comments interspersed in comments.

```sh
ls \
  # comment
  continued
```

(not completely true, interspersed comments will mess with vim's autoindentation but not the syntax highlighting with `sh` filetype).

> ```

Reviewer 3: > Neither GitHub nor Vim correctly highlight comment continuations. POSIXly correct but not correctly implemented.
> However neither GH nor Vim has a problem with comments interspersed in comments.

Huh, same for Emacs :-)

> Why would use those over just prepending `#` to the continued line?

I probably wouldn't, I was just arguing that deviating from standard sh could be surprising to users and implementors. IMHO comments on line continuations are not worth that deviation.

> ```

Author: > IMHO comments on line continuations are not worth that deviation.

I think they are. but maybe @cneud or @EEngl52 can be tiebreaker here - this detail should not prevent us from moving forward and I'm fine with scratching that feature if I'm alone in my stand against POSIX line continuation strictness :)

> ```

Reviewer 4: I certainly wouldn't mind and actually for me it is easier to just use `#` for every line

> ```

Author: > I certainly wouldn't mind and actually for me it is easier to just use `#` for every line

My question was more about comments in continued lines, e.g

```
ocrd-foo \
  # documenting this specific parameter
  -P bla 42 \
  # and this input file group
  -I FOO
```

But I see no point of defending this if no one else sees the benefit. I'll drop that as an official "feature". As the parsing algorithm is defined at the moment, it would still work though, this needs to be aligned with POSIX then.

should begin with `#!/usr/bin/env ocrd-wf-v1`.

### Comments

Any line that begins with a number of whitespace (tab `\x09` or space `\x20`) characters
(including zero) followed by the character "#" (*pound sign*, `\x23`).

### Line continuation

If a line ends with "\" (*backslash*, `\x5c`), backslash and newline are
removed and the next line, without leading whitespace, is appended. This is
repeated as long as the next line ends with a backslash.

### Variable Assignments

An OCRD-WF can contain variable assignments of the form `name=value\n`. `name` must be
a valid `sh` identifier. Everything to the right of `=` is parsed as a string after expanding quotation (i.e. removing enclosing pairs of double `"` or single `'` quotes, or backslash `\` to _escape_ the special meaning of the following character, with `\\`, `\"`, and `\'` for literal backslash, double quote, and single quote, respectively). Note: whitespace is not allowed anywhere, except as quoted part of `value`.

<!-- TODO define behavior? -->
**NOTE** The semantics of assigning `value` to `name` are implementation-dependent

## Parsing Algorithm

The parsing algorithm consists of a set of passes. It should inform

> implementers, but does not require require an exact implementation, as long as

Reviewer 4: ```suggestion
implementers, but does not require an exact implementation, as long as
```
